spec/wizards/schools/register_ect_wizard/ect_spec.rb

```diff
@@ -322,4 +322,150 @@
+          expect(ect.induction_start_date).to eq(Date.new(2023, 6, 10))
+        end
+      end
+
+      context 'when the teacher has no induction periods' do
+        it 'returns nil' do
+          expect(ect.induction_start_date).to be_nil
+        end
+      end
+    end
+
+    describe '#previous_appropriate_body_name' do
+      context 'when the teacher has induction periods' do
+        let!(:older_body) { FactoryBot.create(:appropriate_body, name: 'Older Body') }
+        let!(:more_recent_body) { FactoryBot.create(:appropriate_body, name: 'More Recent Body') }
+
+        before do
+          FactoryBot.create(:induction_period, teacher:, started_on: Date.new(2023, 6, 10), finished_on: Date.new(2023, 9, 30), appropriate_body: older_body)
+          FactoryBot.create(:induction_period, teacher:, started_on: Date.new(2023, 10, 1), finished_on: Date.new(2024, 4, 30), appropriate_body: more_recent_body)
+        end
+
+        it 'returns the name of the latest appropriate body by started_on' do
+          expect(ect.previous_appropriate_body_name).to eq('More Recent Body')
+        end
+      end
+
+      context 'when the teacher has no induction periods' do
+        it 'returns nil' do
+          expect(ect.previous_appropriate_body_name).to be_nil
+        end
+      end
+    end
+
+    describe '#previous_training_programme' do
+      context 'when the teacher has ECTAtSchoolPeriods' do
+        before do
+          FactoryBot.create(:ect_at_school_period, teacher:, programme_type: :school_led, lead_provider_id: nil, started_on: Date.new(2023, 10, 1), finished_on: Date.new(2023, 12, 1))
+          FactoryBot.create(:ect_at_school_period, teacher:, programme_type: :provider_led, started_on: Date.new(2024, 1, 1), finished_on: Date.new(2024, 6, 1))
+        end
+
+        it 'returns the programme type from the latest ECTAtSchoolPeriod by started_on' do
+          expect(ect.previous_training_programme).to eq('provider_led')
+        end
+      end
+
+      context 'when the teacher has no ECTAtSchoolPeriods' do
+        it 'returns nil' do
+          expect(ect.previous_training_programme).to be_nil
+        end
+      end
+    end
+
+    describe '#previous_provider_led?' do
+      it 'returns true when the latest ECTAtSchoolPeriod is provider-led' do
+        FactoryBot.create(:ect_at_school_period, teacher:, programme_type: :provider_led, started_on: Date.new(2024, 1, 1), finished_on: Date.new(2024, 6, 1))
+        expect(ect.previous_provider_led?).to be(true)
+      end
+
+      it 'returns false when the latest ECTAtSchoolPeriod is school-led' do
+        FactoryBot.create(:ect_at_school_period, teacher:, programme_type: :school_led, lead_provider_id: nil, started_on: Date.new(2024, 1, 1), finished_on: Date.new(2024, 6, 1))
+        expect(ect.previous_provider_led?).to be(false)
+      end
+
+      it 'returns nil when there are no ECTAtSchoolPeriods' do
+        expect(ect.previous_provider_led?).to be_nil
+      end
+    end
```

Comment:
I always like to convert the `...when...` side in a context where you prepare the test case and then the `it` (test case) inside but surely is a personal preference. It is true that sometimes too much nested blocks is annoying.

Reply:
I have refactored into context blocks, I think it is slightly nicer too. But yeah, you're right, too many nested context blocks can get confusing. I think there is a Rubocop rule that steers people away from it.